doc/design.md

```diff
@@ -0,0 +1,1268 @@
+
+
+## <a name="design-components"></a>Design components
+
+The following is only an incomplete description.
+For more details and examples the prototype code serves as a reference.
+A good starting point that gives an overview of how `Dataset` may be *used* is given by [`TableWorkspace` example](../test/TableWorkspace_test.cpp), [`Workspace2D` example](../test/Workspace2D_test.cpp), and [`EventWorkspace` example](../test/EventWorkspace_test.cpp).
+
+
+### <a name="design-components-dataset"></a>`class Dataset`
+
+`class Dataset` is literally just a set or list of variables.
+- Variables can be inserted and removed.
+- Iteration over variables is possible.
+- Variables are identified by a combination of their tag and (for data variables) their name.
+  Duplicate tag/name combinations are forbidden, i.e., `Dataset` behaves like a set.
+
+The current implementation also contains a `Dimensions` objects, holding the space spanned by the combination of all the variables dimensions.
+However, this is not necessary and may change in the actual implementation.
+
+There are very few methods for direct manipulation of `Dataset`:
+- `setSlice` to set a slice of the dataset.
+
+The current implementation in the prototype is mostly complete regarding the actual `Dataset` interface.
+Most other functionality would probably be added as free functions.
+- Some interface cleanup for finding variables or checking for existence of certain tags/names is required.
+- `merge` is currently a member method, but should probably be made a free function.
+  Given the copy-on-write mechanism there is no advantage in having it as a method.
+- Insertion mechanism for bin-edge variables needs to be clarified.
+
+
+### <a name="design-components-variable"></a>`class Variable`
+
+`class Variable` can hold a N-dimensional array of arbitrary data, with some additions that make it self-describing:
+- Tag and name identifying the variable.
+- Unit of the data held by the variable.
+- Dimensions of the data held by the variable, i.e., the side lengths of a multi-dimensional cube of data, using `class Dimensions`.
+- Data, stored in an array-like container.
+
+In the current implementation, as much as possible has been moved out of the header to reduce compile times due to templated code.
+The interface implemented currently should be mostly complete with little change required for the final implementation.
+
+##### Type erasure
+
+The type erasure and implementation of basic operations for `Variable` uses **concept-based polymorphism** (a pattern by Sean Parent).
+The involved types are `class VariableConcept` (the interface) and `template <class T> class VariableModel` (the implementation(s)).
+
+##### Alignment
+
+In the prototype we align the storage of data in `Variable` to support vectorization.
+There is no disadvantage to this apart from preventing move operations when initialized from a `std::vector`.
```

Comment:
I think aligned storage is the right thing to do, but I'm wondering if we also need a sanity check whether these vector instructions are compatible with our current use of `NaN` for masking. For example, I know adding `-ffast-math` breaks a bunch of unit tests.

See for example, https://blogs.msdn.microsoft.com/vcblog/2015/10/19/do-you-prefer-fast-or-precise/

Reply:
Thanks for the link. Some googling inspired by your comment also brought up this interesting overview by the famous Agner Fog: https://www.agner.org/optimize/nan_propagation.pdf.